python/ray/tune/examples/mnist_tune_serve.py

```diff
@@ -0,0 +1,498 @@
+        ])
+        self.dataset = MNIST(
+            self.data_dir, train=True, download=True, transform=transform)
+
+    def _get_day_slice(self, day=0):
+        if day < 0:
+            return 0
+        n = len(self.dataset)
+        # Start with 30% of the data, get more data each day
+        return ceil(n * (0.3 + 0.7 * day / self.max_days))
+
+    def get_data(self, day=0):
+        """Get complete normalized train and validation data to date."""
+        end = self._get_day_slice(day)
+
+        available_data = Subset(self.dataset, list(range(end)))
+        train_n = int(0.8 * end)  # 80% train data, 20% validation data
+
+        return random_split(available_data, [train_n, end - train_n])
+
+    def get_incremental_data(self, day=0):
+        """Get next normalized train and validation data day slice."""
+        start = self._get_day_slice(day - 1)
+        end = self._get_day_slice(day)
+
+        available_data = Subset(self.dataset, list(range(start, end)))
+        train_n = int(
+            0.8 * (end - start))  # 80% train data, 20% validation data
+
+        return random_split(available_data, [train_n, end - start - train_n])
+
+
+class ConvNet(nn.Module):
+    def __init__(self, layer_size=192):
+        super(ConvNet, self).__init__()
+        self.layer_size = layer_size
+        self.conv1 = nn.Conv2d(1, 3, kernel_size=3)
+        self.fc = nn.Linear(192, self.layer_size)
+        self.out = nn.Linear(self.layer_size, 10)
+
+    def forward(self, x):
+        x = F.relu(F.max_pool2d(self.conv1(x), 3))
+        x = x.view(-1, 192)
+        x = self.fc(x)
+        x = self.out(x)
+        return F.log_softmax(x, dim=1)
+
+
+def train(model, optimizer, train_loader, device=None):
+    device = device or torch.device("cpu")
+    model.train()
+    for batch_idx, (data, target) in enumerate(train_loader):
+        data, target = data.to(device), target.to(device)
+        optimizer.zero_grad()
+        output = model(data)
+        loss = F.nll_loss(output, target)
+        loss.backward()
+        optimizer.step()
+
+
+def test(model, data_loader, device=None):
+    device = device or torch.device("cpu")
+    model.eval()
+    correct = 0
+    total = 0
+    with torch.no_grad():
+        for batch_idx, (data, target) in enumerate(data_loader):
+            data, target = data.to(device), target.to(device)
+            outputs = model(data)
+            _, predicted = torch.max(outputs.data, 1)
+            total += target.size(0)
+            correct += (predicted == target).sum().item()
+
+    return correct / total
+
+
+# Define Ray Serve model,
+class MNISTBackend:
+    def __init__(self, checkpoint_dir, config, metrics):
+        self.checkpoint_dir = checkpoint_dir
+        self.config = config
+        self.metrics = metrics
+
+        # use_cuda = config.get("use_gpu") and torch.cuda.is_available()
+        # self.device = torch.device("cuda" if use_cuda else "cpu")
```

Comment:
any reason this is commented out?

Reply:
to run on CI iiuc

Reply:
Thanks for catching this. We'll just go with the fully commented `tune-serve-integration-mnist.py` which is the same code. There, this part is not commented out.